mockld/callback_helpers.go

```diff
@@ -0,0 +1,65 @@
+package mockld
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/launchdarkly/sdk-test-harness/framework"
+
+	"github.com/gorilla/mux"
+)
+
+// callbackService provides support infrastructure for callback services that simulate an SDK
+// component with multiple methods. The test service will call one of several predefined subpaths
+// for each component method. For simplicity, and to make it clear that these calls should never
+// be cached, the method is always POST, except for a DELETE method to the base path which means
+// the SDK has stopped using the component. The parameters and responses, if any, are always JSON.
+type callbackService struct {
+	router *mux.Router
+	logger framework.Logger
+	name   string
+}
+
+func newCallbackService(logger framework.Logger, name string) *callbackService {
+	router := mux.NewRouter()
+	c := &callbackService{router: router, logger: logger, name: name}
+	router.HandleFunc("/", c.close).Methods("DELETE")
+	return c
+}
+
+func (c *callbackService) addPath(path string, handler func(*json.Decoder) (interface{}, error)) {
+	c.router.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+		var requestDecoder *json.Decoder
+		var body []byte
+		if r.Body != nil {
+			body, _ = ioutil.ReadAll(r.Body)
+			_ = r.Body.Close()
+			requestDecoder = json.NewDecoder(bytes.NewBuffer(body))
+		}
+		c.logger.Printf("[%s] got POST %s %s", c.name, path, string(body))
```

Comment:
It could be useful to surround the body format string with brackets or something to make it clear to people reading the logs the absence of a body.

Reply:
Well, the reason I didn't bother doing something like that was, the presence or absence of a request body is just part of the service API spec— that is, a type of request that has parameters will always have them, and a type of request that doesn't have parameters will never have a body— that is not a thing that varies according to some variable SDK condition. In other words, if I'm looking at the log and I see a request was made to `/getMetadata`, I'm not going to be wondering "I don't see a request body, was there one?" because there's no reason for `/getMetadata` to have one.